### PR TITLE
cpp multicore impl

### PR DESCRIPTION
**Pull request by Trisfald**
`complete defeat`

### CODE REVIEW

**Comment by LesnyRumcajs** [comment]
does it impact single-core performance?

**Comment by Trisfald** [comment]
yes, it obliterates single core performance

**Comment by LesnyRumcajs** [comment]
> yes, it obliterates single core performance

Shouldn't we have it then in a separate benchmark?

**Comment by Trisfald** [comment]
> > yes, it obliterates single core performance
> 
> Shouldn't we have it then in a separate benchmark?

Not a bad idea actually

File: build.sh [hunk 1]
```diff
@@ -8,6 +8,8 @@ docker build --force-rm -f go_grpc/Dockerfile -t go_grpc_test .
 
 docker build --force-rm -f cpp_grpc/Dockerfile -t cpp_grpc_test .
```
Comment (LesnyRumcajs) [inline build.sh:9]: should we call it `mt` for clarity?

File: cpp_grpc/main.cpp [hunk 2]
```diff
@@ -60,13 +66,20 @@ class ServerImpl final {
     builder.RegisterService(&service_);
     // Get hold of the completion queue used for the asynchronous communication
     // with the gRPC runtime.
-    cq_ = builder.AddCompletionQueue();
+    auto parallelism = std::max(1u, std::thread::hardware_concurrency() / 4);
```
Comment (LesnyRumcajs) [inline cpp_grpc/main.cpp:69]: 💪

File: build.sh (hunk 1 above)
Comment (Trisfald) [reply build.sh:9]: for now the convention is

`x_test` : multithread
`x_st_test`: specialized st version

but we can have either

`x_test` : generic all round rocksolid impl
OR
`x_mt_test`: specialized mt version
PLUS
`x_st_test`: specialized st version

File: cpp_grpc/main.cpp (hunk 2 above)
Comment (Trisfald) [reply cpp_grpc/main.cpp:69]: this's so bad I'm ashamed

File: build.sh (hunk 1 above)
Comment (LesnyRumcajs) [reply build.sh:9]: I think the latter seems closer to reality and more expressive - Go impl seems generic enough and scales nicely with more threads while C++ / Rust does not. Additionally just looking at the names would suggest which technologies need additional tweaking to achieve desired performance.

**Pull request merged**